lib/useKResponsiveWindow.js

```diff
@@ -0,0 +1,199 @@
+
+  /**
+   * @function addMediaQuery
+   */
+  const orientationQuery = addMediaQuery(`(orientation: portrait)`, event => {
+    updateOrientation(event.matches);
+    updateWindow();
+    updateGutter();
+  });
+
+  /**
+   * @function addMediaQuery
+   */
+  const heightQuery = addMediaQuery(`(max-height: 600px)`, event => {
+    windowIsShort.value = event.matches;
+    updateWindow();
+  });
+
+  /**
+   * Generate width media queries
+   */
+  const widthQueries = (() => {
+    const SCROLL_BAR = 16;
+    const queries = [
+      '(max-width: 480px)',
+      '(max-width: 600px)',
+      '(max-width: 840px)',
+      `(max-width: ${960 - SCROLL_BAR}px)`,
+      `(max-width: ${1280 - SCROLL_BAR}px)`,
+      `(max-width: ${1440 - SCROLL_BAR}px)`,
+      `(max-width: ${1600 - SCROLL_BAR}px)`,
+      `(min-width: 1601px)`,
+    ];
+
+    const mediaQueries = [];
+    queries.forEach((query, index) => {
+      const widthQuery = addMediaQuery(query, () => {
+        windowBreakpoint.value = index;
+        updateWindow();
+        updateGutter();
+      });
+      mediaQueries.push(widthQuery);
+    });
+    return mediaQueries;
+  })();
+
+  /**
+   * Initialize properties
+   */
+  const initProps = () => {
+    orientationQuery.eventHandler(orientationQuery.mediaQueryList);
+    heightQuery.eventHandler(heightQuery.mediaQueryList);
+    widthQueries.every(widthQuery => {
+      if (widthQuery.mediaQueryList.matches) {
+        widthQuery.eventHandler(widthQuery.mediaQueryList);
+        return false;
+      }
+      return true;
+    });
+  };
+
+  /**
+   * Stop listening to media query changes
+   */
+  const stopListening = () => {
+    orientationQuery.stopListening();
+    heightQuery.stopListening();
+    widthQueries.forEach(mediaQuery => {
+      const { stopListening } = mediaQuery;
+      stopListening();
+    });
+  };
+
+  /**
+   * Update window gutter
+   */
+  const updateGutter = () => {
+    if (windowIsSmall.value) {
+      windowGutter.value = 16;
+    } else if (smallestWindowDimensionIsLessThan(600)) {
+      windowGutter.value = 16;
+    } else {
+      windowGutter.value = 24;
+    }
+  };
+
+  /**
+   * Check if the smallest window dimension(width or height) is smaller than the specified dimension
+   * @param {Number} dimension in px
+   * @returns {Boolean}
+   */
+  const smallestWindowDimensionIsLessThan = dimension => {
+    return (
+      windowBreakpoint.value < 4 && Math.min(windowWidth.value, windowHeight.value) < dimension
+    );
+  };
+
+  /**
+   * Update window orientation
+   * @param {Boolean} portrait
+   */
+  const updateOrientation = portrait => {
+    windowIsPortrait.value = portrait;
+    windowIsLandscape.value = !windowIsPortrait.value;
+  };
+
+  /**
+   * Update window width and height
+   */
+  const updateWindow = () => {
+    const metrics = windowMetrics();
+    windowWidth.value = metrics.width;
+    windowHeight.value = metrics.height;
+  };
```

Comment:
The only potential issue that I see moving towards the media query approach here is that there may be usages relying on width and height being accurate, regardless of breakpoints. From looking through your code, it seems the width and height will only update when transitioning between breakpoints and during orientation changes. Is that correct? I guess this could be something to audit in Kolibri when you integrate this work. Perhaps those should leverage something more specifically tailored to their behavior instead of this composable.

Reply:
This could mean that this composable doesn't even return `windowWidth` and `windowHeight`, and instead there's an explicit composable that could be used for that to avoid unnecessary listeners on window resizing.

Reply:
One way to retain previous behaviour (but still providing an advantage over the mixin approach) is to have a single module scoped event listener listening to the window resize event, that is also updating a set of module scoped reactive data references for tracking the window width and height.

This way we only ever need one global listener, and a single reactive source of truth for the current window dimensions.

@MisRob and I have used this approach of module scoped reactives combined with function scoped reactives in order to provide some global elements and some component specific elements, e.g. https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/learn/assets/src/composables/useLearnerResources.js#L22

Reply:
The other advantage here, is then the existing mixin could just transparently use the logic in this module in its `setup` method, and then the mixin just becomes a compatibility layer so that we don't have to migrate everything at once.

Reply:
@bjester your interpretation is correct. As suggested by @rtibbles, we need something better tailored for dimension changes. The onresize listeners on the mixin seem to do the trick. Will proceed as suggested

Reply:
@rtibbles The existing [mixin](https://github.com/learningequality/kolibri-design-system/blob/develop/lib/KResponsiveWindowMixin.js#L79) already had a single global listener on the window's `resize`. Ideally that should also be throttled through `requestAnimationFrame` too.

I think a separate composable makes sense for the resize event handling, tracking width and height, since that  functionality that should be deprecated going forward. And we definitely shouldn't be using that for recreating the wheel that already exists in `matchMedia`!

Reply:
Right, the main thing we could do here is only have a single function that gets invoked on the resize event, as we only need to update a single module level reactive, rather needing to update the `height` and `width` on every single active mixin, so the number of handler functions invoked no longer scales with the number of uses of the mixin.

I agree that implementing this as a separate composable as a compatibility layer would be useful, so we can deprecate and then remove it, once we migrate all uses of the mixin to no longer directly rely on `height` and `width` (I think this shouldn't be too many, as most should be referring to the breakpoints).

Reply:
a new composable `useWindowDimensions` has been created and is being called in `lib/useKResponsiveWindow/index.js`